src/lib/operators/set_union.cpp

```diff
@@ -0,0 +1,371 @@
+ * performing the SetUnion  operation.
+ * Since sorting a multi-column matrix by rows would require a lot of value-copying, for each ReferenceMatrix, a
+ * VirtualPosList is created.
+ * Each element of this VirtualPosList references a row in a ReferenceMatrix by index. This way, if two values need to
+ * be swapped while sorting, only two indices need to swapped instead of a RowIDs for each column in the
+ * ReferenceMatrices.
+ * Using a implementation derived from std::set_union, the two virtual pos lists are merged into the result table.
+ *
+ *
+ * ### About ReferenceMatrices
+ * The ReferenceMatrix consists of N rows and C columns of RowIDs.
+ * N is the same number as the number of rows in the input table.
+ * Each of the C column can represent 1..S columns in the input table. All rows represented by a ReferenceMatrix-Column
+ * contain the same PosList in each Chunk.
+ *
+ * The ReferenceMatrix of a StoredTable will only contain one column, the ReferenceMatrix of the result of a 3 way Join
+ * will contain 3 columns
+ *
+ *
+ * ### TODO(anybody) for performance improvements
+ * Instead of using a ReferenceMatrix, consider using a linked list of RowIDs for each row. Since most of the sorting
+ *      will depend on the leftmost column, this way most of the time no remote memory would need to be accessed
+ *
+ * The sorting, which is the most expensive part of this operator, could probably be parallelized.
+ */
+
+namespace {
+
+// See doc above
+using ReferenceMatrix = std::vector<opossum::PosList>;
+
+using VirtualPosList = std::vector<size_t>;
+
+/**
+ * Comparator for performing the std::sort() of a virtual pos list
+ */
+struct VirtualPosListCmpContext {
+  ReferenceMatrix& reference_matrix;
+  bool operator()(size_t lhs, size_t rhs) const {
+    for (const auto& reference_matrix_column : reference_matrix) {
+      const auto left_row_id = reference_matrix_column[lhs];
+      const auto right_row_id = reference_matrix_column[rhs];
+
+      if (left_row_id < right_row_id) return true;
+      if (right_row_id < left_row_id) return false;
+    }
+    return false;
+  }
+};
+
+}  // namespace
+
+namespace opossum {
+
+SetUnion::SetUnion(const std::shared_ptr<const AbstractOperator>& left,
+                   const std::shared_ptr<const AbstractOperator>& right)
+    : AbstractReadOnlyOperator(left, right) {}
+
+uint8_t SetUnion::num_in_tables() const { return 2; }
+
+uint8_t SetUnion::num_out_tables() const { return 1; }
+
+std::shared_ptr<AbstractOperator> SetUnion::recreate(const std::vector<AllParameterVariant>& args) const {
+  return std::make_shared<SetUnion>(input_left()->recreate(args), input_right()->recreate(args));
+}
+
+const std::string SetUnion::name() const { return "SetUnion"; }
+
+const std::string SetUnion::description() const { return "SetUnion"; }
+
+std::shared_ptr<const Table> SetUnion::_on_execute() {
+  const auto early_result = _analyze_input();
```

Comment:
This is a weird "does a lot of things"-method. It validates the input, it checks for early results, it fills `_column_segment_begins`, and more.

Reply:
explained _prepare_operator() better. Can't really split this up, would become something like
```
_validate_input_a();
_prepare_operator();
_validate_input_b()
```

since some of the validation depends on the prepare step. I'll leave it as it is.